### PR TITLE
Update search match rules

### PR DESCRIPTION
**Pull request by ReaNAiveD**
## Update Fuzzy search word match rule
- Match exactly if word length <= `EXACT_MATCH_LENGTH` (default is 3)
- Use fuzzy distance 1 if word length <= `DISTANCE_1_MATCH_LENGTH` (default is 5)
- Use fuzzy distance 2 if word length > `DISTANCE_1_MATCH_LENGTH`

## Update multi keywords match rule
- Use new parameter `match_type` to control `and/or` behavior

## `top_num` limitation
- Limit `top_num` to 1-20

### CODE REVIEW

File: API/SearchService/__init__.py [hunk 1]
```diff
@@ -14,12 +14,18 @@ def main(req: func.HttpRequest,
 
     try:
         keyword = get_param_str(req, "keyword", required=True)
-        search_type = get_param_search_type(req, "type", default=SearchType.All)
+        search_type = get_param_search_type(
+            req, "type", default=SearchType.All)
```
Comment (zhoxing-ms) [inline API/SearchService/__init__.py]:
```suggestion
        search_type = get_param_search_type(req, "type", default=SearchType.All)
```

File: API/SearchService/__init__.py [hunk 2]
```diff
@@ -16,10 +16,15 @@ def main(req: func.HttpRequest,
...
+        if top_num <=0 or top_num > 20:
+            raise ParameterException("Illegal parameter: the parameter 'top_num' must be in the range 1-20")
+        match_type = get_param_match_type(req, "match_type", default=MatchType.All)
     except ParameterException as e:
         return func.HttpResponse(e.msg, status_code=400)
-    results = get_search_results(
-        build_search_statement(keyword), top_num, search_type.get_search_fields())
+    results = get_search_results(build_search_statement(keyword, match_type), top_num, search_type.get_search_fields())
+    if len(keyword.split()) > 1 and len(results) < top_num and match_type == MatchType.All:
+        or_results = get_search_results(build_or_search_statement(keyword), top_num, search_type.get_search_fields())
```
Comment (zhoxing-ms) [inline API/SearchService/__init__.py:26]:
```suggestion
        or_results = get_search_results(build_or_search_statement(keyword), top_num-len(results), search_type.get_search_fields())
```

Comment (ReaNAiveD) [reply API/SearchService/__init__.py:26]: There can be some duplicate search results in `and` search and `or` search. So I don't think it's appropriate to limit extra search `top_num`.

File: API/SearchService/src/util.py [hunk 3]
```diff
@@ -71,5 +78,63 @@ def get_param_search_type(req: func.HttpRequest, name: str, required=False, defa
...
+    dist1_length = os.environ.get("DISTANCE_1_MATCH_LENGTH", 4)
+    for word in keyword.split():
+        if not word:
+            continue
+        if len(word) <= exact_length:
+            word = word
+        elif len(word) <= dist1_length:
+            word = word + "~1"
+        else:
+            word = word + "~"
```
Comment (zhoxing-ms) [inline API/SearchService/src/util.py]: The explicit declaration of distance will make the code more readable, so we don't pay attention to the default distance represented by `~`
```suggestion
            word = word + "~2"
```

File: API/SearchService/src/util.py [hunk 4]
```diff
@@ -71,5 +78,63 @@ def get_param_search_type(req: func.HttpRequest, name: str, required=False, defa
...
+
+
+def build_or_search_statement(keyword: str) -> str:
+    return _build_search_statement(keyword, join_word=" OR ")
+
+
+def _build_search_statement(keyword: str, join_word=" AND ") -> str:
+    search_statement = []
+    exact_length = os.environ.get("EXACT_MATCH_LENGTH", 2)
+    dist1_length = os.environ.get("DISTANCE_1_MATCH_LENGTH", 4)
```
Comment (zhoxing-ms) [inline API/SearchService/src/util.py]: Personally, I don't think there will be typo when the word length is less than or equal to 3
```suggestion
    exact_length = os.environ.get("EXACT_MATCH_LENGTH", 3)
    dist1_length = os.environ.get("DISTANCE_1_MATCH_LENGTH", 5)
```

File: API/SearchService/__init__.py (hunk 2 above)
Comment (zhoxing-ms) [reply API/SearchService/__init__.py:26]: Got it, thanks~

Comment (zhoxing-ms) [reply API/SearchService/__init__.py:26]: @ReaNAiveD However, do we have logic to limit the number of returned results according to `top_num`? 
If not, is it possible that the number of results we return will be greater than `top_num`?

Comment (ReaNAiveD) [reply API/SearchService/__init__.py:26]: Sorry. I forget to put a limit on merged results.

Comment (ReaNAiveD) [reply API/SearchService/__init__.py:26]: We should add 
```
results = results[:min(top_num, len(results))]
```
after next line `append_results(results, or_results)`.

File: API/SearchService/__init__.py [hunk 5]
```diff
@@ -16,10 +16,16 @@ def main(req: func.HttpRequest,
...
+        append_results(results, or_results)
+        results = results[:min(top_num, len(results))]
```
Comment (zhoxing-ms) [inline API/SearchService/__init__.py]:
```suggestion
        results = results[:top_num]
```

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**